sky/utils/command_runner.py

```diff
@@ -275,14 +278,23 @@ def run(
                 command += [f'> {log_path}']
             executable = '/bin/bash'
 
-        return log_lib.run_with_log(' '.join(command),
-                                    log_path,
-                                    require_outputs=require_outputs,
-                                    stream_logs=stream_logs,
-                                    process_stream=process_stream,
-                                    shell=True,
-                                    executable=executable,
-                                    **kwargs)
+        backoff = common_utils.Backoff(initial_backoff=5, max_backoff_factor=5)
+        result = None
+        while max_retry >= 0:
+            result = log_lib.run_with_log(' '.join(command),
+                                          log_path,
+                                          require_outputs=require_outputs,
+                                          stream_logs=stream_logs,
+                                          process_stream=process_stream,
+                                          shell=True,
+                                          executable=executable,
+                                          **kwargs)
+            returncode = result[0] if require_outputs else result
+            if returncode == 0:
```

Comment:
To clarify, do you mean that we should not retry for codes 1, 2, 127? When do these occur? I'm thinking that we instead of just breaking on 0, we should pass in a set of `end_return_codes`, the set of return codes for which to stop retrying. We would then set this depending on the command.

In the case of a timeout for autostop (255), I believe that we should retry, since it may be due to a transient network issue.

Reply:
> do you mean that we should not retry for codes 1, 2, 127? When do these occur?

It's meant to be a question to investigate :) 


For 255, you can reproduce as follows. Launch a VM, manually terminate it, then `ssh vm`. Eventually it'll throw 255 after a while. So it's not just for transient. In this case, is there any `sky` call paths that can lead to `run()` being called on this VM? If so, then retrying would prolong the wait.

Reply:
From `ssh` man page: 
```
ssh exits with the exit status of the remote command or with 255 if an error occurred.
``` 
`run()` always runs the given `cmd` using `ssh`. Therefore, the meaning of the return codes (1, 2, 127) depends on the actual command executed. So, when specifying `max_retry`, we should update `end_return_codes` so that the function stops retrying correctly based on context.

Specifically for `set_autostop`, since the command executed is `python` (which returns 0 by default), the default of stopping on 0 works. Other return codes do not apply here since we never override the default return value in `autostop_lib`.

For a 255 return code, indeed it does not mean that there is a transient error, but there is no way of distinguishing between transient errors and cases where the user kills the VM beyond the return code. Since users frequently face autostop failing, I think the tradeoff of longer wait time is OK in this case. The extra time required for an extra SSH command will also not be very long if the VM is already down.